File: src/main/java/dev/isxander/xanderlib/utils/Multithreading.java
package dev.isxander.xanderlib.utils;

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public final class Multithreading {

    private static final AtomicInteger counter = new AtomicInteger(0);
    private static final ScheduledThreadPoolExecutor SCHEDULED_POOL = new ScheduledThreadPoolExecutor(20, (r) -> new Thread(r, String.format("Scheduled Thread #%s", counter.incrementAndGet())));
    public static final ThreadPoolExecutor POOL = new ThreadPoolExecutor(20, 20, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), (r) -> new Thread(r, String.format("Thread #%s", counter.incrementAndGet())));

    public static ScheduledFuture<?> scheduleAsyncAtFixedRate(Runnable runnable, long initialDelay, long delayBetweenExecution, TimeUnit unit) {
        return SCHEDULED_POOL.scheduleAtFixedRate(runnable, initialDelay, delayBetweenExecution, unit);
    }

    public static ScheduledFuture<?> scheduleAsync(Runnable runnable, long delay, TimeUnit unit) {
        return SCHEDULED_POOL.schedule(runnable, delay, unit);
    }

    public static void runAsync(Runnable runnable) {
        POOL.execute(runnable);
    }

    public static Future<?> submit(Runnable runnable) {
        return POOL.submit(runnable);
    }

}

File: src/main/kotlin/dev/isxander/xanderlib/utils/Multithreading.kt
package dev.isxander.xanderlib.utils

import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicInteger

object Multithreading {

    private val counter = AtomicInteger(0)
    private val SCHEDULED_POOL = ScheduledThreadPoolExecutor(
        20
    ) { r: Runnable? ->
        Thread(
            r,
            "Scheduled Thread ${counter.incrementAndGet()}"
        )
    }
    private val POOL = ThreadPoolExecutor(
        20, 20, 0L, TimeUnit.SECONDS, LinkedBlockingQueue()
    ) { r: Runnable? -> Thread(r, String.format("Thread #%s", counter.incrementAndGet())) }

    fun scheduleAsync(runnable: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture<*>? {
        return SCHEDULED_POOL.schedule(runnable, delay, unit)
    }

    fun runAsync(runnable: Runnable) {
        POOL.execute(runnable)
    }

}